// dynamodb-toolbox@0.3.5/dist/__tests__/bootstrap-tests.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocumentClient2 = exports.delay = exports.DocumentClient = exports.DynamoDB = exports.dynaliteServer = exports.createTableParams = void 0;
// Include dynalite and config server (use in memory)
// @ts-ignore: No type file available
const dynalite_1 = __importDefault(require("dynalite"));
// Load simple table creation parameters
exports.createTableParams = require('./tables/create-table.json');
exports.dynaliteServer = dynalite_1.default({
    createTableMs: 0,
    updateTableMs: 0,
    deleteTableMs: 0
});
// Load AWS SDK
const aws_sdk_1 = __importDefault(require("aws-sdk"));
// Create DynamoDB connection to dynalite
exports.DynamoDB = new aws_sdk_1.default.DynamoDB({
    endpoint: 'http://localhost:4567',
    region: 'us-east-1',
    credentials: new aws_sdk_1.default.Credentials({
        accessKeyId: 'test',
        secretAccessKey: 'test'
    })
});
// Create our document client
exports.DocumentClient = new aws_sdk_1.default.DynamoDB.DocumentClient({
    endpoint: 'http://localhost:4567',
    region: 'us-east-1',
    credentials: new aws_sdk_1.default.Credentials({
        accessKeyId: 'test',
        secretAccessKey: 'test'
    })
    // convertEmptyValues: true
});
// Delay helper
const delay = (ms) => new Promise(res => setTimeout(res, ms));
exports.delay = delay;
exports.DocumentClient2 = new aws_sdk_1.default.DynamoDB.DocumentClient({
    region: 'us-east-1',
    credentials: new aws_sdk_1.default.SharedIniFileCredentials({ profile: '' })
    // convertEmptyValues: false
});
